Content.Server/GameObjects/Components/NodeNetwork/INodeComponent.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Content.Server.GameObjects.Components.NodeGroup
+{
+    /// <summary>
+    ///     Organizes themselves into distinct <see cref="INodeGroup"/>s with other <see cref="INodeComponent"/>s
+    ///     that they can "reach" and have the same <see cref="NodeGroupID"/>.
+    /// </summary>
+    public interface INodeComponent
+    {
+        NodeGroupID NodeGroupID { get; }
+
+        public INodeGroup NodeGroup { get; set; }
+
+        bool TryAssignGroupIfNeeded();
+
+        void SpreadRemadeGroup();
+    }
+
+    public abstract class NodeComponent : Component, INodeComponent
+    {
+        [ViewVariables]
+        public NodeGroupID NodeGroupID => _nodeGroupID;
+        private NodeGroupID _nodeGroupID;
+
+        [ViewVariables]
+        public INodeGroup NodeGroup { get => _nodeGroup; set => SetNodeGroup(value); }
+        private INodeGroup _nodeGroup;
+
+#pragma warning disable 649
+        [Dependency] private readonly INodeGroupFactory _nodeGroupFactory;
+#pragma warning restore 649
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _nodeGroupID, "nodeGroupID", NodeGroupID.Default);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            TryAssignGroupIfNeeded();
+            SpreadGroup();
+        }
+
+        public override void OnRemove()
+        {
+            base.OnRemove();
+            NodeGroup.RemoveNode(this);
+            _nodeGroup = null;
+        }
+
+        public bool TryAssignGroupIfNeeded()
+        {
+            if (NodeGroup != null)
+            {
+                return false;
+            }
+            NodeGroup = GetReachableCompatibleGroups().FirstOrDefault() ?? MakeNewGroup();
+            return true;
+        }
+
+        public void SpreadRemadeGroup()
+        {
+            Debug.Assert(NodeGroup != null);
+            foreach (var node in GetReachableCompatibleNodes().Where(node => node.NodeGroup == null))
+            {
+                node.NodeGroup = NodeGroup;
+                node.SpreadRemadeGroup();
+            }
+        }
+
+        /// <summary>
+        ///     Strategy for how to find other reachable <see cref="INodeComponent"/>s to group with.
+        /// </summary>
+        protected abstract IEnumerable<NodeComponent> GetReachableNodes();
+
+        private IEnumerable<NodeComponent> GetReachableCompatibleNodes()
+        {
+            return GetReachableNodes().Where(node => node.NodeGroupID == NodeGroupID)
+                .Where(node => node != this);
```

Comment:
This seems unnecessary, given that the docs for `GetReachableNodes` specifies "other". I think it should be clarified if `GetReachableNodes` should return the current node, since otherwise someone is going to forget this check somewhere.